src/server/ua_services_nodemanagement.c

```diff
@@ -203,11 +203,13 @@ typeCheckVariableNode(UA_Server *server, UA_Session *session,
     }
 
     /* Check array dimensions against the vt */
-    if(!compatibleArrayDimensions(vt->arrayDimensionsSize, vt->arrayDimensions,
```

Comment:
What does the change do?
If the VariableType explicitly requires a certain arrayDimensions, then we always should test for that.

Reply:
-2 is "scalar or an array with any number of dimensions". The OPC UA foundation nodesets seem to rely on that. Without this exception, several nodes can't be instantiated. You can try this out by only applying the changes for the python scripts and starting a server with generated full namespace 0, DI, PLCOpen and ADI  nodesets.

Reply:
In that case, the fault is with the VariableType for the new node.

If the VariableType explicitly requires an arrayDimension, how can the ValueRank be -2?

Reply:
An example for this is ns=1;i=10620 in the ADI nodeset. It has no value rank specified. From how I understand the spec, this means that the value rank from the type definition must be used.
I have added a commit that fills ValueRank and ArrayDimensions from the type definition in the nodeset compiler. With this, the second condition can be removed.

The "more correct" way to solve the problem would be checking the specifiedAttributes mask when instantiating a node with a type definition in the server.